<a href="https://colab.research.google.com/github/melaniedaniel7/CFPB-Text-Analysis-Using-NLP-Techniques/blob/main/Exploring%20the%20Dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Exploring the Dataset**

## **Objectives:**



* Import the dataset
* Sort the data by product and sub-product
* Draw conclusions
* Sort the proritised data by issue
* Draw conclusions





---



# **Import the dataset**

In [2]:
# Imported the dataset
from google.colab import files

uploaded = files.upload()

Saving complaints-2026-06-17_10_50.csv to complaints-2026-06-17_10_50.csv


In [3]:
# Checked the dataset uploaded correctly
import pandas as pd

df = pd.read_csv("complaints-2026-06-17_10_50.csv")
df.head()
df.columns

Index(['Date received', 'Product', 'Sub-product', 'Issue', 'Sub-issue',
       'Consumer complaint narrative', 'Company public response', 'Company',
       'State', 'ZIP code', 'Tags', 'Submitted via', 'Date sent to company',
       'Company response to consumer', 'Timely response?', 'Complaint ID'],
      dtype='object')

In [4]:
# Checked if any complaint narratives were missing
df.shape
df['Consumer complaint narrative'].isnull().sum()

np.int64(0)

# **Sort the data by product and sub-product**

In [5]:
# Sort the data

# Sorting by product and subproduct
# From filtering the dataset on the CFPB website, I already knew there were 4 product categories
# This code was created with the help of Google Gemini, which is built into Google Colab
grouped_products = df.groupby('Product')['Sub-product'].agg(['nunique', lambda x: x.unique().tolist()])
grouped_products.columns = ['Number of Sub-products', 'List of Sub-products']

# Reset index to make 'Product' a column and add 'Product number'
grouped_products = grouped_products.reset_index()
grouped_products.insert(0, 'Product number', range(1, 1 + len(grouped_products)))

display(grouped_products)

,Product number,Product,Number of Sub-products,List of Sub-products
0,1,Credit card,2,"[General-purpose credit card or charge card, S..."
1,2,Debt collection,11,"[I do not know, Rental debt, Telecommunication..."
2,3,Mortgage,8,"[FHA mortgage, Conventional home mortgage, VA ..."
3,4,Student loan,2,"[Federal student loan servicing, Private stude..."


In [6]:
# Seeing how many consumer complaints there are for each product and sub-product
# This code was created with the help of Google Gemini, which is built into Google Colab
complaints_per_product_subproduct = df.groupby(['Product', 'Sub-product']).size().reset_index(name='Number of Complaints')
display(complaints_per_product_subproduct)

,Product,Sub-product,Number of Complaints
0,Credit card,General-purpose credit card or charge card,3884
1,Credit card,Store credit card,577
2,Debt collection,Auto debt,261
3,Debt collection,Credit card debt,1360
4,Debt collection,Federal student loan debt,23
5,Debt collection,I do not know,3342
6,Debt collection,Medical debt,330
7,Debt collection,Mortgage debt,26
8,Debt collection,Other debt,1325
9,Debt collection,Payday loan debt,110


# **Draw conclusions**

From these results, I decided to prioritise the dataset and consumer complaints by picking a single subcategory for each product with the highest number of complaints.

My thought process here was to try to find a way to ensure that the identified topics are as accurate and reliable as possible.

This is because in phase two I identified the most prevalent topics across the corpus but as one can see from the table above this would most likely not be an accurate reflection of the data and the consumer comlaints.

For example, within the corpus, there are a total of 4461 consumer complaints for the product "Credit card," but 759 consumer complaints for the product "Student loan".

Therefore, I will only be experimenting with the following data moving forward:

1) Credit card (General-purpose credit card or charge card) = 3884 complaints
2) Debt collection (I do not know) = 3342 complaints
3) Mortgage (Conventional home mortgage)= 866 complaints
4) Student loan (Federal student loan servicing) = 594 complaints

# **Sort the proritised data by issue**

In [16]:
# Here I used Gemini to help me turn each of the prioritised data groups above into its own dataset.

# Filter for Credit card (General-purpose credit card or charge card)
credit_card_df = df[(df['Product'] == 'Credit card') &
                    (df['Sub-product'] == 'General-purpose credit card or charge card')].copy()

# Filter for Debt collection (I do not know)
debt_collection_df = df[(df['Product'] == 'Debt collection') &
                        (df['Sub-product'] == 'I do not know')].copy()

# Filter for Mortgage (Conventional home mortgage)
mortgage_df = df[(df['Product'] == 'Mortgage') &
                 (df['Sub-product'] == 'Conventional home mortgage')].copy()

# Filter for Student loan (Federal student loan servicing)
student_loan_df = df[(df['Product'] == 'Student loan') &
                     (df['Sub-product'] == 'Federal student loan servicing')].copy()

print(f"Credit card (General-purpose credit card or charge card) complaints: {len(credit_card_df)}")
display(credit_card_df.head())

print(f"Debt collection (I do not know) complaints: {len(debt_collection_df)}")
display(debt_collection_df.head())

print(f"Mortgage (Conventional home mortgage) complaints: {len(mortgage_df)}")
display(mortgage_df.head())

print(f"Student loan (Federal student loan servicing) complaints: {len(student_loan_df)}")
display(student_loan_df.head())

Credit card (General-purpose credit card or charge card) complaints: 3884


,Date received,Product,Sub-product,Issue,Sub-issue,Consumer complaint narrative,Company public response,Company,State,ZIP code,Tags,Submitted via,Date sent to company,Company response to consumer,Timely response?,Complaint ID
0,2026-04-21T18:37:38.000Z,Credit card,General-purpose credit card or charge card,"Other features, terms, or problems",Privacy issues,I was told by XXXX XXXX that XXXX XXXX XXXX g...,Company believes it acted appropriately as aut...,Atlanticus Services Corporation,MS,39120,NaN,Web,2026-04-21T18:51:13.000Z,Closed with explanation,Yes,21451090
2,2026-04-21T20:02:43.000Z,Credit card,General-purpose credit card or charge card,Problem with a purchase shown on your statement,Card was charged for something you did not pur...,"To Whom It May Concern, I am submitting this c...",Company believes it acted appropriately as aut...,Atlanticus Services Corporation,VA,23832,NaN,Web,2026-04-21T20:24:22.000Z,Closed with non-monetary relief,Yes,21455712
5,2026-04-21T23:00:18.000Z,Credit card,General-purpose credit card or charge card,"Other features, terms, or problems",Other problem,Bank of America is refusing to execute a XXXX ...,Company has responded to the consumer and the ...,"BANK OF AMERICA, NATIONAL ASSOCIATION",NC,28715,NaN,Web,2026-04-21T23:04:18.000Z,Closed with explanation,Yes,21462779
12,2026-04-22T12:18:00.000Z,Credit card,General-purpose credit card or charge card,Problem with a purchase shown on your statement,Credit card company isn't resolving a dispute ...,DEMAND LETTER OFFICE OF THE PRESIDENT SYNCHRON...,Company has responded to the consumer and the ...,SYNCHRONY FINANCIAL,NY,10031,NaN,Web,2026-04-22T12:35:22.000Z,Closed with explanation,Yes,21478535
16,2026-04-22T17:08:47.000Z,Credit card,General-purpose credit card or charge card,Trouble using your card,Can't use card to make purchases,I opened a small business credit card with Ban...,Company has responded to the consumer and the ...,"BANK OF AMERICA, NATIONAL ASSOCIATION",NV,89509,NaN,Web,2026-04-22T17:32:59.000Z,Closed with non-monetary relief,Yes,21490000


Debt collection (I do not know) complaints: 3342


,Date received,Product,Sub-product,Issue,Sub-issue,Consumer complaint narrative,Company public response,Company,State,ZIP code,Tags,Submitted via,Date sent to company,Company response to consumer,Timely response?,Complaint ID
1,2026-04-21T19:50:01.000Z,Debt collection,I do not know,Electronic communications,Frequent or repeated messages,I have been receiving text messages with a hyp...,Company disputes the facts presented in the co...,"Eastern Account Systems of Connecticut, Inc.",CT,06320,NaN,Web,2026-04-21T19:57:43.000Z,Closed with explanation,Yes,21454377
4,2026-04-21T22:51:23.000Z,Debt collection,I do not know,False statements or representation,Attempted to collect wrong amount,XXXX XXXX XXXX XXXX XXXX XXXX XXXX NC XXXX XX/...,Company believes it acted appropriately as aut...,"Professional Finance Company, Inc.",NC,28269,NaN,Web,2026-04-21T23:23:46.000Z,Closed with explanation,Yes,21463484
11,2026-04-22T09:23:55.000Z,Debt collection,I do not know,False statements or representation,Attempted to collect wrong amount,( CREDIT CONTROL LLC ) This is a formal reques...,Company has responded to the consumer and the ...,"Credit Control, LLC",GA,31705,NaN,Web,2026-04-22T09:28:54.000Z,Closed with explanation,Yes,21476201
15,2026-04-22T15:47:08.000Z,Debt collection,I do not know,Attempts to collect debt not owed,Debt is not yours,Disputed Account Information Account Name : CC...,Company can't verify or dispute the facts in t...,"CCS Financial Services, Inc.",FL,33025,NaN,Web,2026-04-22T15:51:10.000Z,Closed with explanation,Yes,21485298
42,2026-04-24T14:17:44.000Z,Debt collection,I do not know,Attempts to collect debt not owed,Debt is not yours,I am requesting validation of the alleged debt...,Company has responded to the consumer and the ...,WELLS FARGO & COMPANY,TX,77385,NaN,Web,2026-04-24T14:21:15.000Z,Closed with non-monetary relief,Yes,21564795


Mortgage (Conventional home mortgage) complaints: 866


,Date received,Product,Sub-product,Issue,Sub-issue,Consumer complaint narrative,Company public response,Company,State,ZIP code,Tags,Submitted via,Date sent to company,Company response to consumer,Timely response?,Complaint ID
6,2026-04-22T00:20:22.000Z,Mortgage,Conventional home mortgage,Trouble during payment process,Trying to communicate with the company to fix ...,CFPB CONSUMER COMPLAINT : IDENTITY FRAUD & PRO...,Company has responded to the consumer and the ...,WELLS FARGO & COMPANY,GA,30064,NaN,Web,2026-04-22T00:26:14.000Z,Closed with explanation,Yes,21465417
7,2026-04-22T00:27:52.000Z,Mortgage,Conventional home mortgage,Trouble during payment process,Trying to communicate with the company to fix ...,CFPB CONSUMER COMPLAINT : CIRCULAR FRAUD & UNC...,Company has responded to the consumer and the ...,WELLS FARGO & COMPANY,GA,30064,NaN,Web,2026-04-22T00:33:48.000Z,Closed with explanation,Yes,21465588
8,2026-04-22T01:16:15.000Z,Mortgage,Conventional home mortgage,Trouble during payment process,Payment process,"I received a full accounting from Shellpoint, ...",Company has responded to the consumer and the ...,"Shellpoint Partners, LLC",MD,20879,NaN,Web,2026-04-22T01:51:44.000Z,Closed with explanation,Yes,21467703
20,2026-04-23T00:56:34.000Z,Mortgage,Conventional home mortgage,Trouble during payment process,"Escrow, taxes, or insurance",When Wells Fargo answered to CFPB they lied to...,Company has responded to the consumer and the ...,WELLS FARGO & COMPANY,CA,95687,"Older American, Servicemember",Web,2026-04-23T01:36:15.000Z,Closed with explanation,Yes,21509736
37,2026-04-24T05:06:48.000Z,Mortgage,Conventional home mortgage,Struggling to pay mortgage,Foreclosure,This complaint presents new evidence two IRS F...,Company has responded to the consumer and the ...,"BANK OF AMERICA, NATIONAL ASSOCIATION",FL,33610,NaN,Web,2026-04-24T05:41:44.000Z,Closed with explanation,Yes,21555757


Student loan (Federal student loan servicing) complaints: 594


,Date received,Product,Sub-product,Issue,Sub-issue,Consumer complaint narrative,Company public response,Company,State,ZIP code,Tags,Submitted via,Date sent to company,Company response to consumer,Timely response?,Complaint ID
79,2026-03-17T15:59:52.000Z,Student loan,Federal student loan servicing,Struggling to repay your loan,Can't get other flexible options for repaying ...,I am filing a complaint against my federal stu...,Company believes it acted appropriately as aut...,SLM CORPORATION,MD,20912,NaN,Web,2026-03-17T16:08:50.000Z,Closed with explanation,Yes,20335936
132,2026-04-07T01:39:37.000Z,Student loan,Federal student loan servicing,Dealing with your lender or servicer,Problem with customer service,I have been fighting for over XXXX years now f...,NaN,MOHELA,GA,30281,Servicemember,Web,2026-04-07T01:51:20.000Z,Untimely response,No,20996476
134,2026-04-07T11:20:52.000Z,Student loan,Federal student loan servicing,Dealing with your lender or servicer,Problem with customer service,I am submitting a complaint because technical ...,NaN,Servicer under contract with Federal Student Aid,NaN,XXXXX,Older American,Web,2026-04-07T11:47:31.000Z,Untimely response,No,21005739
136,2026-04-07T13:42:25.000Z,Student loan,Federal student loan servicing,Incorrect information on your report,Account information incorrect,Request a review and correction of student loa...,NaN,"Nelnet, Inc.",FL,32757,NaN,Web,2026-04-07T14:12:20.000Z,Closed with explanation,Yes,21009046
141,2026-04-07T15:04:18.000Z,Student loan,Federal student loan servicing,Dealing with your lender or servicer,Problem with customer service,I have provided evidence that these loans were...,NaN,"Nelnet, Inc.",PA,18974,Older American,Web,2026-04-07T15:49:55.000Z,Closed with explanation,Yes,21012929


In [15]:
# Sort the data
# Here I wanted to explore the issues and sub-issues of the prioritised data
# Here I once again used Google Gemini

# Issues and Sub-issues for Credit Card (General-purpose credit card or charge card)
print("\n--- Issues and Sub-issues for Credit Card (General-purpose credit card or charge card) ---")
grouped_credit_card_issues = credit_card_df.groupby('Issue')['Sub-issue'].agg(['nunique', lambda x: x.unique().tolist()])
grouped_credit_card_issues.columns = ['Number of Sub-issues', 'List of Sub-issues']
print(f"Total Number of Issues: {len(grouped_credit_card_issues)}")
display(grouped_credit_card_issues.reset_index())


--- Issues and Sub-issues for Credit Card (General-purpose credit card or charge card) ---
Total Number of Issues: 14


,Issue,Number of Sub-issues,List of Sub-issues
0,"Advertising and marketing, including promotion...",2,[Didn't receive advertised or promotional term...
1,Closing your account,2,"[Company closed your account, Can't close your..."
2,Credit monitoring or identity theft protection...,5,"[Billing dispute for services, Didn't receive ..."
3,Fees or interest,3,"[Problem with fees, Charged too much interest,..."
4,Getting a credit card,5,"[Card opened without my consent or knowledge, ..."
5,Improper use of your report,3,[Reporting company used your report improperly...
6,Incorrect information on your report,6,"[Information belongs to someone else, Account ..."
7,"Other features, terms, or problems",8,"[Privacy issues, Other problem, Problem with b..."
8,Problem when making payments,2,"[Problem during payment process, You never rec..."
9,Problem with a company's investigation into an...,5,[Their investigation did not fix an error on y...


In [17]:
print("\n--- Number of Complaints per Issue for Credit Card (General-purpose credit card or charge card) ---")
credit_card_issues_complaint_count = credit_card_df.groupby('Issue').size().reset_index(name='Number of Complaints')
display(credit_card_issues_complaint_count.sort_values(by='Number of Complaints', ascending=False))


--- Number of Complaints per Issue for Credit Card (General-purpose credit card or charge card) ---


,Issue,Number of Complaints
10,Problem with a purchase shown on your statement,1468
7,"Other features, terms, or problems",508
3,Fees or interest,366
4,Getting a credit card,354
1,Closing your account,261
8,Problem when making payments,243
0,"Advertising and marketing, including promotion...",198
13,Trouble using your card,179
6,Incorrect information on your report,141
9,Problem with a company's investigation into an...,84


In [18]:
# Issues and Sub-issues for Debt Collection (I do not know)
print("\n--- Issues and Sub-issues for Debt Collection (I do not know) ---")
grouped_debt_collection_issues = debt_collection_df.groupby('Issue')['Sub-issue'].agg(['nunique', lambda x: x.unique().tolist()])
grouped_debt_collection_issues.columns = ['Number of Sub-issues', 'List of Sub-issues']
print(f"Total Number of Issues: {len(grouped_debt_collection_issues)}")
display(grouped_debt_collection_issues.reset_index())


--- Issues and Sub-issues for Debt Collection (I do not know) ---
Total Number of Issues: 7


,Issue,Number of Sub-issues,List of Sub-issues
0,Attempts to collect debt not owed,4,"[Debt is not yours, Debt was result of identit..."
1,Communication tactics,4,"[You told them to stop contacting you, but the..."
2,Electronic communications,4,"[Frequent or repeated messages, You told them ..."
3,False statements or representation,4,"[Attempted to collect wrong amount, Told you n..."
4,Threatened to contact someone or share informa...,3,"[Contacted your employer, Talked to a third-pa..."
5,Took or threatened to take negative or legal a...,7,[Threatened or suggested your credit would be ...
6,Written notification about debt,3,[Didn't receive enough information to verify d...


In [19]:
print("\n--- Number of Complaints per Issue for Debt Collection (I do not know) ---")
debt_collection_issues_complaint_count = debt_collection_df.groupby('Issue').size().reset_index(name='Number of Complaints')
display(debt_collection_issues_complaint_count.sort_values(by='Number of Complaints', ascending=False))


--- Number of Complaints per Issue for Debt Collection (I do not know) ---


,Issue,Number of Complaints
0,Attempts to collect debt not owed,2111
6,Written notification about debt,416
3,False statements or representation,408
5,Took or threatened to take negative or legal a...,265
1,Communication tactics,98
2,Electronic communications,27
4,Threatened to contact someone or share informa...,17


In [20]:
# Issues and Sub-issues for Mortgage (Conventional home mortgage)
print("\n--- Issues and Sub-issues for Mortgage (Conventional home mortgage) ---")
grouped_mortgage_issues = mortgage_df.groupby('Issue')['Sub-issue'].agg(['nunique', lambda x: x.unique().tolist()])
grouped_mortgage_issues.columns = ['Number of Sub-issues', 'List of Sub-issues']
print(f"Total Number of Issues: {len(grouped_mortgage_issues)}")
display(grouped_mortgage_issues.reset_index())


--- Issues and Sub-issues for Mortgage (Conventional home mortgage) ---
Total Number of Issues: 7


,Issue,Number of Sub-issues,List of Sub-issues
0,Applying for a mortgage or refinancing an exis...,8,"[Application denials, Changes in loan terms du..."
1,Closing on a mortgage,5,[Closing disclosure or other related disclosur...
2,Improper use of your report,1,[Reporting company used your report improperly]
3,Incorrect information on your report,4,"[Account status incorrect, Information belongs..."
4,Problem with a company's investigation into an...,4,"[Problem with personal statement of dispute, T..."
5,Struggling to pay mortgage,4,"[Foreclosure, Trying to communicate with the c..."
6,Trouble during payment process,9,[Trying to communicate with the company to fix...


In [21]:
print("\n--- Number of Complaints per Issue for Mortgage (Conventional home mortgage) ---")
mortgage_issues_complaint_count = mortgage_df.groupby('Issue').size().reset_index(name='Number of Complaints')
display(mortgage_issues_complaint_count.sort_values(by='Number of Complaints', ascending=False))


--- Number of Complaints per Issue for Mortgage (Conventional home mortgage) ---


,Issue,Number of Complaints
6,Trouble during payment process,492
5,Struggling to pay mortgage,200
0,Applying for a mortgage or refinancing an exis...,78
1,Closing on a mortgage,57
3,Incorrect information on your report,22
4,Problem with a company's investigation into an...,16
2,Improper use of your report,1


In [22]:
# Issues and Sub-issues for Student loan (Federal student loan servicing)
print("\n--- Issues and Sub-issues for Student loan (Federal student loan servicing) ---")
grouped_student_loan_issues = student_loan_df.groupby('Issue')['Sub-issue'].agg(['nunique', lambda x: x.unique().tolist()])
grouped_student_loan_issues.columns = ['Number of Sub-issues', 'List of Sub-issues']
print(f"Total Number of Issues: {len(grouped_student_loan_issues)}")
display(grouped_student_loan_issues.reset_index())


--- Issues and Sub-issues for Student loan (Federal student loan servicing) ---
Total Number of Issues: 7


,Issue,Number of Sub-issues,List of Sub-issues
0,Credit monitoring or identity theft protection...,2,"[Billing dispute for services, Problem with pr..."
1,Dealing with your lender or servicer,7,"[Problem with customer service, Trouble with h..."
2,Improper use of your report,2,[Credit inquiries on your report that you don'...
3,Incorrect information on your report,7,"[Account information incorrect, Information be..."
4,Problem with a company's investigation into an...,4,[Their investigation did not fix an error on y...
5,Problem with fraud alerts or security freezes,0,[nan]
6,Struggling to repay your loan,6,[Can't get other flexible options for repaying...


In [23]:
print("\n--- Number of Complaints per Issue for Student loan (Federal student loan servicing) ---")
student_loan_issues_complaint_count = student_loan_df.groupby('Issue').size().reset_index(name='Number of Complaints')
display(student_loan_issues_complaint_count.sort_values(by='Number of Complaints', ascending=False))


--- Number of Complaints per Issue for Student loan (Federal student loan servicing) ---


,Issue,Number of Complaints
1,Dealing with your lender or servicer,355
6,Struggling to repay your loan,130
3,Incorrect information on your report,74
4,Problem with a company's investigation into an...,19
2,Improper use of your report,11
0,Credit monitoring or identity theft protection...,4
5,Problem with fraud alerts or security freezes,1


# **Draw conclusions**

Here I was able to see what issues made up the majority of the consumer complaint groups I prioritised for exploration.

The data shows that for:
1) Credit card (General-purpose credit card or charge card), the majority of complaints (1468) were related to "Problem with a purchase shown on your statement"
2) Debt collection (I do not know), the majority of complaints (2111) were related to "Attempts to collect debt not owed"
3) Mortgage (Conventional home mortgage), the majority of complaints (492) were related to "Trouble during payment process"
4) Student loan (Federal student loan servicing), the majority of complaints (355) were related to "Dealing with your lender or servicer"

Although this information was insightful I decided not to filter the data further using these issue categories.

Additionally, I decided that filtering the consumer complaints by these other categories may defeat the purpose of this project. Therefore, I will only use this data again at the end of the project when I reflect on the most prevalent topics identified.